File: 0981-time-based-key-value-store/0981-time-based-key-value-store.py
class TimeMap:

    def __init__(self):
        self.mp = {}

    def set(self, key: str, value: str, timestamp: int) -> None:
        if not key in self.mp:
            self.mp[key] = []
        self.mp[key].append((value,timestamp))

    def get(self, key: str, timestamp: int) -> str:
        if not key in self.mp:
            return ""
        lst = self.mp[key]
        l   = 0
        r   = len(lst)-1
        idx = -1
        while l <= r:
            mid = (l+r)//2
            if lst[mid][1] <= timestamp:
                l   = mid+1
                idx = mid
            else:
                r = mid-1
        return "" if idx == -1 else lst[idx][0]


# Your TimeMap object will be instantiated and called as such:
# obj = TimeMap()
# obj.set(key,value,timestamp)
# param_2 = obj.get(key,timestamp)

File: 0981-time-based-key-value-store/0981-time-based-key-value-store.cpp
class TimeMap {
public:
    unordered_map<string,vector<pair<int,string>>> mp;

    TimeMap() {
    }
    
    void set(string key, string value, int timestamp) {
        mp[key].push_back({timestamp,value});
    }
    
    string get(string key, int timestamp) {
        if (!mp.contains(key))
            return "";
        auto& vec = mp[key];
        int n = (int)vec.size();
        int l = 0;
        int r = n-1;
        int idx = -1;
        while (l <= r){
            int mid = l + (r-l) / 2;
            if (vec[mid].first <= timestamp){
                idx = mid;
                l   = mid+1;
            } else 
                r   = mid-1;
        }
        return (idx == -1) ? "" : vec[idx].second;
    }
};

/**
 * Your TimeMap object will be instantiated and called as such:
 * TimeMap* obj = new TimeMap();
 * obj->set(key,value,timestamp);
 * string param_2 = obj->get(key,timestamp);
 */